price-feeder/oracle/provider/huobi.go

```diff
@@ -27,140 +32,248 @@ type (
 	//
 	// REF: https://huobiapi.github.io/docs/spot/v1/en/#market-data
 	HuobiProvider struct {
-		baseURL string
-		client  *http.Client
-	}
-
-	// HuobiTraceResponse defines the response type for the last executed trade
-	// for a given ticker/symbol.
-	HuobiTraceResponse struct {
-		Status string `json:"status"`
-		ErrMsg string `json:"err-msg"`
-		Tick   struct {
-			Data []struct {
-				Price float64 `json:"price"`
-			} `json:"data"`
-		} `json:"tick"`
-	}
-
-	// HuobiMarketResponse defines the response type for the last 24h market summary
-	// for a given ticker/symbol.
-	HuobiMarketResponse struct {
-		Status string `json:"status"`
-		ErrMsg string `json:"err-msg"`
-		Tick   struct {
-			Vol float64 `json:"vol"`
-		} `json:"tick"`
+		wsURL           url.URL
+		wsClient        *websocket.Conn
+		logger          zerolog.Logger
+		mu              sync.Mutex
+		tickers         map[string]HuobiTicker // market.$symbol.ticker => HuobiTicker
+		subscribedPairs []types.CurrencyPair
+		reconnectTicker *time.Ticker
+	}
+
+	// HuobiTicker defines the response type for the channel and
+	// the tick object for a given ticker/symbol
+	HuobiTicker struct {
+		CH   string    `json:"ch"` // Data belonged channel，Format：market.$symbol.ticker
+		Tick HuobiTick `json:"tick"`
+	}
+
+	// HuobiTick defines the response type for the last 24h market summary
+	// and the last traded price for a given ticker/symbol
+	HuobiTick struct {
+		Vol       float64 `json:"vol"`       // Accumulated trading value of last 24 hours
+		LastPrice float64 `json:"lastPrice"` // Last traded price
+	}
+
+	// HuobiSubscriptionMsg Msg to subscribe to one ticker channel at time
+	HuobiSubscriptionMsg struct {
+		Sub string `json:"sub"` // channel to subscribe market.$symbol.ticker
 	}
 )
 
-func NewHuobiProvider() *HuobiProvider {
-	return &HuobiProvider{
-		baseURL: huobiBaseURL,
-		client:  newDefaultHTTPClient(),
+func NewHuobiProvider(ctx context.Context, logger zerolog.Logger, pairs ...types.CurrencyPair) (*HuobiProvider, error) {
+	wsURL := url.URL{
+		Scheme: "wss",
+		Host:   huobiHost,
+		Path:   huobiPath,
 	}
-}
 
-func NewHuobiProviderWithTimeout(timeout time.Duration) *HuobiProvider {
-	return &HuobiProvider{
-		baseURL: huobiBaseURL,
-		client:  newHTTPClientWithTimeout(timeout),
+	wsConn, _, err := websocket.DefaultDialer.Dial(wsURL.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("error connecting to Huobi websocket: %w", err)
 	}
+
+	provider := &HuobiProvider{
+		wsURL:           wsURL,
+		wsClient:        wsConn,
+		logger:          logger.With().Str("module", "oracle").Logger(),
+		tickers:         map[string]HuobiTicker{},
+		reconnectTicker: time.NewTicker(huobiReconnectTime),
+	}
+
+	if err := provider.subscribeTickers(pairs...); err != nil {
+		return nil, err
+	}
+	provider.subscribedPairs = pairs
```

Comment:
why not have this be part of the struct literal creation on line 76?

Reply:
You can update the other providers too :)

Reply:
done at https://github.com/umee-network/umee/pull/569/commits/9c7f0c6af5cf96bbafdb3373fe52aca44e51b2b7